File: Problem581/Solution.go
package main

func findUnsortedSubarray(nums []int) int {
	var end int
	var start int
	var max int
	start = len(nums)
	end = -1
	max = nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] < max {
			end = getMax(end, i)
			position := i - 1
			num := nums[i]
			for position >= 0 && num < nums[position] {
				position--
			}
			start = getMin(start, position)
		} else {
			max = nums[i]
		}

	}
	if end == -1 {
		return 0
	}
	return end - start
}

func getMax(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func getMin(a int, b int) int {
	if a > b {
		return b
	}
	return a
}

File: Problem581/Solution.java
/**
 * @author tgorthi
 * @since Jun 2020
 */
public class Solution
{
    public int findUnsortedSubarray(int[] nums)
    {
        int start = Integer.MAX_VALUE;
        int end = -1;
        int max = nums[0];
        for (int i = 1; i < nums.length; i++)
        {
            if (nums[i] < max)
            {
                end = Math.max(end, i);
                int tmp = i - 1;
                int num = nums[i];
                while (tmp >= 0 && num < nums[tmp])
                {
                    tmp--;
                }
                start = Math.min(start, tmp + 1);
            }
            else
            {
                max = nums[i];
            }
        }
        return end == -1 ? 0 : end - start + 1;
    }
}